File: go/perfect-numbers/perfect_numbers.go
package perfect

import "errors"

type Classification string

const (
	ClassificationDeficient Classification = "Classification Deficient"
    ClassificationPerfect Classification = "Classification Perfect"
    ClassificationAbundant Classification = "Classification Abundant"
)

var ErrOnlyPositive = errors.New("input is not positive")

func Classify(n int64) (Classification, error) {
    if n <= 0 {
        return "", ErrOnlyPositive
    }
    aliquotSum := int64(0)
	for i := int64(1); i < n; i++ {
        if n % i == 0 {
            aliquotSum += i
        }
    }
	if aliquotSum > n {
        return ClassificationAbundant, nil
    } else if aliquotSum == n {
    	return ClassificationPerfect, nil
    } else {
    	return ClassificationDeficient, nil
    }
}

File: cpp/perfect-numbers/perfect_numbers.cpp
#include "perfect_numbers.h"
#include <stdexcept>

namespace perfect_numbers {
    Classification classify(int num) {
        if (num < 1) throw std::domain_error("input must be positive");
        int aliquot_sum{0};
        for (int i = 1; i <= num / 2; i++) {
            if (num % i == 0) aliquot_sum += i;
        }
        if (aliquot_sum == num) return perfect;
        else if (aliquot_sum < num) return deficient;
        else return abundant;
    }
}  // namespace perfect_numbers
